bitvm/src/hash/blake3_u4_compact.rs

```diff
@@ -0,0 +1,617 @@
+                    }
+
+                    for _ in 0..(4-j) * 2{
+                        OP_DROP
+                    }
+
+                    for _ in 0..(4-j) * 2{
+                        OP_0
+                    }
+
+                    for _ in 0..(j*2){
+                        OP_FROMALTSTACK
+                    }
+
+                    for _ in 0..(pad_bytes*2){
+                        OP_0
+                    }
+                ),
+                0,
+                false,
+                0,
+                "padding",
+            );
+        }
+
+        //make a hashmap of msgs
+        let mut original_message = Vec::new();
+        for i in 0..16 {
+            let m = stack.define(8, &format!("msg_{}", i));
+            original_message.push(m);
+        }
+
+        // create the current block message map
+        let mut message = HashMap::new();
+        for m in 0..16 {
+            message.insert(m as u8, original_message[m as usize]);
+        }
+
+        compress(
+            stack,
+            i != 0,
+            0,
+            msg_len.min(64),
+            get_flags_for_block(i, num_blocks),
+            message,
+            &tables,
+            8,
+            i == num_blocks - 1,
+        );
+
+        //delete the intermediate states
+        for _ in 0..8 {
+            stack.drop(stack.get_var_from_stack(0));
+        }
+
+        if msg_len > 64 {
+            msg_len -= 64;
+        }
+    }
+    // drop tables
+    tables.drop(stack);
+
+    // get the result hash
+    stack.from_altstack_joined(8 as u32 * 8, "blake3-hash");
+}
+
+#[cfg(any(fuzzing, test))]
+//verifies that the hash of the input byte slice matches with official implementation.
+pub fn test_blake3_compact_givenbyteslice(input_bytes: &[u8]) -> String{
+    use rand::random;
+
+
+    let mut stack = StackTracker::new();
+
+    let msg_len = input_bytes.len();
+    
+    //compute the official hash
+    let expected_hash = blake3::hash(input_bytes);
+
+
+    //determine amount of padding needed to get the compact working
+    let padding_bytes_needed = if msg_len % 64 == 0 {0} else {64 - (msg_len % 64)};
+    let mut padded_msg = input_bytes.to_vec();
+    padded_msg.extend(std::iter::repeat(0u8).take(padding_bytes_needed));
+
+    assert!(padded_msg.len() % 64 == 0, "padding failed");
+
+       // reverse the 4 byte chunks
+    for chunk in padded_msg.chunks_mut(4) {
+        chunk.reverse();
+    }
+ 
+
+    // push the msg into the stack and compact it
+    stack.custom(script!(
+        for chunk in padded_msg.chunks(64).rev(){
+            for (i,byte) in chunk.iter().enumerate(){
+                {*byte}
+                if i == 31 || i == 63{
+                    {U256::transform_limbsize(8,29)}
+                }
+            }
+        } 
+    ),
+    0,
+    false,
+    0,
+    "push_msgs"
+    );
+
+    //random bool to decide if we want to use full table or half table.
+    let use_full_table_toogle: bool = random();
```

Comment:
Recommend making this a parameter for the function so the caller can control it more directly.